### ISSUE TITLE
Dedicated docs site

### ISSUE THREAD

**Issue by haveyaseen** [redux-orm/redux-orm#261]
We should create a dedicated website that goes to greater depth than the Readme on GitHub.
What should we include?

- Simple and minimal API guide
    - Schema definition using Model classes
    - ORM creation and model registration
    - Reducer creation
    - Selector creation with selector specs
- Detailed explanations
    - Overview diagram [#203]
    - Links to relevant code section or JSDoc page
    - Example with React-Redux and relationship selectors 
    - Temporary solution for custom fields as in https://github.com/redux-orm/redux-orm/issues/258#issuecomment-493606448
- Technical requirements
    - `Map`, `Set` and `Symbol` polyfills (already in Readme)
    - JavaScriptCore version requirements for React-Native [#259]

**Comment by markerikson** [comment]
Sounds great to me!  I recommend basing it off of the Docusaurus setup we have for Redux, React-Redux, and Redux Starter Kit.

**Comment by haveyaseen** [comment]
Thanks for pointing this out! We should probably go with Docusaurus :+1:

**Comment by tomasz-zablocki** [comment]
I can pick up this one up, at least part of it. 

I can start with transplantation of Redux-Starter-Kit Docosaurus setup.

**Comment by haveyaseen** [comment]
Any help is much appreciated :+1:  Giving you push access, feel free to act as you wish.

**Comment by ericleite** [comment]
May I also suggest adding a "Best Practices" section to point out ways to optimize performance, and the do's and don'ts for using the various methods like `update` vs `upsert` and `session` vs `mutableSession`?

I think it would also be nice to have an example of how to integrate redux-orm with API calls. It looks like the existing docs assume you are only working with local data, but most practical use cases involve fetching data from an API.

I just got started using this library and I am loving it so far, but the existing examples reference different versions of the redux-orm API and I'm not sure what is the right way to do things as of today. I'd be happy to help out with documenting these sections as well after I get more familiar with how to use this library.

Just for reference, here are the sources of documentation and usage examples I've come across:
https://redux-orm.github.io/redux-orm/index.html
https://github.com/redux-orm/redux-orm
https://github.com/tommikaikkonen/redux-orm-primer
https://blog.isquaredsoftware.com/2016/10/practical-redux-part-1-redux-orm-basics/

**Comment by tomasz-zablocki** [comment]
I've been messing around Docosaurus today and it seems to be pretty straightforward and robust solution.

Following @markerikson's advice, I've based it on the redux-starter-kit docs structure. 
The goal is to come up with a solid documentation framework to continuously build upon.

So far I've managed to get it working, removed the redux-starter-kit docs and got a placeholder structure, based loosely upon; 

> * Simple and minimal API guide
>   
>   * Schema definition using Model classes
>   * ORM creation and model registration
>   * Reducer creation
>   * Selector creation with selector specs
> * Detailed explanations
>   
>   * Overview diagram [#203]
>   * Links to relevant code section or JSDoc page
>   * Example with React-Redux and relationship selectors
>   * Temporary solution for custom fields as in [#258 (comment)](https://github.com/redux-orm/redux-orm/issues/258#issuecomment-493606448)
> * Technical requirements
>   
>   * `Map`, `Set` and `Symbol` polyfills (already in Readme)
>   * JavaScriptCore version requirements for React-Native [#259]

I've been also playing around the theme a little bit - the violet L&F made it feel like an official Redux team project.

@haveyaseen I'd like to create a branch dedicated to joint documentation efforts and I'll try to deliver the initial Docosaurus structure by the end of the week. 

@ericleite An update to `@types/redux-orm` had been merged, theses are TS flavoured docs, but I've updated several outdated jsdocs comments while writing these. 

On a side note, the spec cases for Redux-ORM are really solid and complete - probably the most up-to-date and valid guide to APIs.

**Comment by tomasz-zablocki** [comment]
Sorry it takes so long, I'm working on it.

Here's a teaser 😉:
  
![teaser](https://user-images.githubusercontent.com/48603138/59149261-a217ef00-8a13-11e9-851a-691f8e2d4ae5.PNG)

@markerikson let me know if I'm too liberal with being inspired by redux doc sites.

**Comment by tomasz-zablocki** [comment]
> Any help is much appreciated 👍 Giving you push access, feel free to act as you wish.

@haveyaseen can you check this push permission ? I'm getting 403 while trying to push my new branch.

**Comment by haveyaseen** [comment]
@tomasz-zablocki You should have received an email with an invite link. I resent it now.

**Comment by tomasz-zablocki** [comment]
@haveyaseen Thank you very much.

First of all, running the docusaurus locally: 

In the website folder:
```bash
npm install
```
and:
```bash
npm run start
```

Docusaurus is really intuitive and it's really easy to get started in minutes.
Here's a short wrap up of changes:
* I had to temporarily change the output destination of existing `build:docs` script, since docusaurus expects the content to be located in `docs` folder.
* `website` folder contains docusaurus configuration, a landing page component and static assets. 
* `docs` folder is where the actual documentation resides. I've added some placeholders to match the proposed structure.
* I've added a gulp task (rough draft) which generates markdown output from sources, using jsdoc-to-markdown, so that the final result is more or less similar to redux/react-redux/redux-starter-kit counterparts.

Now, the last point is something that needs work. 

@markerikson correct me if I'm wrong, but Redux and other Redux library API docs seem to be handcrafted. This would be fine for redux etc. since the public API surface is very small, but not really viable for Redux-ORM.

That output produced by the mentioned gulp task is far from finished, starting with broken links and messed up content in the right sidebar.

Any comments or questions regarding the docs are welcome.

Maybe it would be good to reiterate on the proposed structure (adding more details etc.) and break down missing parts into smaller units to work upon.

**Comment by markerikson** [comment]
Yeah, we've written all our API docs by hand so far.  Not sure what the best approach would be for integrating auto-generated API docs.

**Comment by code-matt** [comment]
As someone who just slogged through learning this over a week, very much needed 😓 The ORM docs themselves are not terrible but hooking it up to redux/saga in a clean way was an internet adventure. Lots of stale or incomplete info out there.


This should go under the Book example in the readme with some explanation:
```
dispatch({ type: 'ORM_ADD_SOME_ENTITIES', data: action.data })
```

```
export default function entitiesReducer(state = orm.getEmptyState(), action) {
  let session;

  // pass the state, action and session to the actual reducers
  switch (true) {
    case Object.values(SOME_MODEL_ACTIONS).includes(action.type):
      session = orm.session(state);
      handleSomeEntitiesActions(state, action, session);
      break;
  }

  // Finally, return state either the default or the updated session.state
  return session ? session.state : state;
}
```

```
export default function (state, action, session) {
  switch (action.type) {
    case 'ORM_ADD_SOME_ENTITIES':
      action.data.forEach(entity => {
        const { stuff, youdonotwant, somearrayofdata, ...restOfValues } = entity
        somearrayofdata.forEach(datum => {
          session.SomeOtherModelWithAnFk.upsert({
            ...datum,
           _someEntityId: rest.id
          })
        })
        session.SomeModel.upsert(restOfValues)
      })
      break

    default:
      break
  }
}

export const SOME_MODEL_ACTIONS = [
  'ORM_ADD_SOME_ENTITIES'
]
```

Thank you @markerikson of course for it.

**Comment by tomasz-zablocki** [comment]
# Docs update

First of all - apologies for the underwhelming progress on this one.

Most of my recent non-commercial coding time had been consumed by updating **redux-orm** typings, answering related questions and solving issues. Practically, I'm the sole author of current type declarations, so I feel responsible for their state. Whenever I had to choose between the types or the docs I ended up with types.

## Auto-generated API reference section

This is almost finished.  There is a bunch of broken links and a few minor formatting issues left. 

I really want this jsdoc->API reference automation to be streamlined and functional since I'm quite confident this is something that will pay off in the long run:

* remove additional effort required to keep the website in sync with sources
* incentivize updating jsdoc to match actual source modifications 
* maybe imposing strict eslint regime on jsdoc content is something worth discussing.

## Concerns regarding remaining sections

I can only speak for myself, but I feel that the issue's scope is way too broad. Whenever I start reasoning about the actual doc content, the reading flow and a host of other things, how detailed a section should be etc, I end up a bit overwhelmed. 
 
I'd like to leverage **projects** feature to break docs related efforts into units with narrower scopes. 

This main goal is to reduce distraction and streamline the process of reasoning about the structure. 
I think it's also safe to assume that tasks with tangible delivery perspective and clear D-o-D have higher chance of attracting contributors.

I'm not going to delude anyone, myself included, that I'm capable of delivering complete docs by myself.  At least not in a time frame that I'd consider satisfying.

**So the first question to @haveyaseen and the rest of the community is: are you fine with me creating a project to organize docs related work on **redux-orm** ?**

I'm confident that by taking an iterative approach, there's a fair better chance that we'll end up with better final result.

**Comment by haveyaseen** [comment]
@tomasz-zablocki Seeing your hard work and your enthusiasm makes me extremely thankful! Sure, if a project seems useful to you, go right ahead and use that feature :+1: Of course our time can be quite limited in this open-source world so anything that helps us prioritize issues is a great win.

The 1.0 roadmap in #140 is already way overbloated which is why I once marked tasks that I deemed most necessary in bold. In general we could translate this to a two-phase backlog: one for the things that need to be worked on ASAP (tangible delivery as you said) and one for things that we really want to do later on. Same for docs related tasks.

I also agree that an iterative strategy is best. Yes, everyone always wants to have things as fast as possible but first of all there's no need to feel pressured! And the first version neither has to be perfect nor complete, as in any case it will outshine the current docs by far :joy: Lately I've been a bit short on time, hopefully in August I can take a closer look at the new docs and help you out.

The auto-generated API reference section is pretty cool. I haven't looked at the visuals yet but I think that generating the docs was so messy before and this might really speed up that process. Last I checked JSDoc templates were really hard to come by as well. All in all, as soon as you think the API section is ready we can merge `feature/standalone-docs` into master, change the GH pages root to the `docs` folder and delete the `gh-pages` branch.

**Comment by haveyaseen** [comment]
@tomasz-zablocki I just added some documentation on the basics of using Redux-ORM (branch `feature/standalone-docs-basics`). Figured you're probably still working on the JSDoc integration so I didn't push into your branch. Due to some error  I had to update the `docusaurus` dependency. But then it worked pretty well and actually it's a joy to work with!

I'm thinking we can add another page to the basics. The first page (called it Quick Start) explains `Model` classes, `ORM` creation and the Redux store integration using `createReducer`. The second explains `Field` definitions. A third one could explain selectors and maybe React-Redux (that's likely what people are going to be using the most?)

**Comment by tomasz-zablocki** [comment]
@haveyaseen Actually for now I'm done with the JSDoc generation. 

Cross-document and local document linking is working now and the` docusaurus` output is 95% sound with correct escaping, expected font styles etc. I'll polish the layout and styling once we have actual JSDoc complete and up-to-date with sources. JSDoc-related todos are gathered in #275, but it's probably not an exhaustive list. There `api-docs:watch` gulp task, that automates the docusaurus update procedure, launching it on every change to `src` or template files. 

Regarding the documentation page content - I'll go through `feature/standalone-docs-basics` now and share my take on structure right afterwards.

**Comment by haveyaseen** [comment]
Closed with #280 :slightly_smiling_face:  An initial version is live at https://redux-orm.github.io/redux-orm/, further contributions are welcome. Thanks for all the hard work!
